constantine/math_compiler/impl_fields_globals.nim

```diff
@@ -100,10 +100,175 @@ func checkValidModulus(M: BigNum) =
+  when T is uint32:
+    const
+      HalfWidth = 16
+      HalfBase = (1'u32 shl HalfWidth)
+      HalfMask = HalfBase - 1
+  elif T is uint64:
+    const
+      HalfWidth = 32
+      HalfBase = (1'u64 shl HalfWidth)
+      HalfMask = HalfBase - 1
+  else:
+    {.error: "Invalid type for BigNum.".}
+
+func hi[T: DynWord](n: T): T =
+  Widths()
+  result = n shr HalfWidth
+
+func lo[T: DynWord](n: T): T =
+  Widths()
+  result = n and HalfMask
+
+func split[T: DynWord](n: T): tuple[hi, lo: T] =
+  result.hi = n.hi
+  result.lo = n.lo
+
+func merge[T: DynWord](hi, lo: T): T =
+  when T is uint32:
+    const HalfWidth = 16
+  else: # uint64
+    const HalfWidth = 32
+  (hi shl HalfWidth) or lo
+
+func addC[T: DynWord](cOut, sum: var T, a, b, cIn: T) =
+  # Add with carry, fallback for the Compile-Time VM
+  # (CarryOut, Sum) <- a + b + CarryIn
+  let (aHi, aLo) = split(a)
+  let (bHi, bLo) = split(b)
+  let tLo = aLo + bLo + cIn
+  let (cLo, rLo) = split(tLo)
+  let tHi = aHi + bHi + cLo
+  let (cHi, rHi) = split(tHi)
+  cOut = cHi
+  sum = merge(rHi, rLo)
+
+func subB[T: DynWord](bOut, diff: var T, a, b, bIn: T) =
+  # Substract with borrow, fallback for the Compile-Time VM
+  # (BorrowOut, Sum) <- a - b - BorrowIn
+  Widths()
+  let (aHi, aLo) = split(a)
+  let (bHi, bLo) = split(b)
+  let tLo = HalfBase + aLo - bLo - bIn
+  let (noBorrowLo, rLo) = split(tLo)
+  let tHi = HalfBase + aHi - bHi - T(noBorrowLo == 0)
+  let (noBorrowHi, rHi) = split(tHi)
+  bOut = T(noBorrowHi == 0)
+  diff = merge(rHi, rLo)
+
+func add[T: DynWord](a: var BigNum[T], w: T): bool =
+  ## Limbs addition, add a number that fits in a word
+  ## Returns the carry
+  var carry, sum: T
+  addC(carry, sum, T(a.limbs[0]), w, carry)
+  a.limbs[0] = sum
+  for i in 1 ..< a.limbs.len:
+    let ai = T(a.limbs[i])
+    addC(carry, sum, ai, 0, carry)
+    a.limbs[i] = sum
+
+  result = bool(carry)
+
+func shiftRight*[T: DynWord](a: var BigNum[T], k: int) =
+  ## Shift right by k.
+  ##
+  ## k MUST be less than the base word size (2^32 or 2^64)
+  const wordBitwidth = sizeof(T) * 8
+  for i in 0 ..< a.limbs.len-1:
+    a.limbs[i] = (a.limbs[i] shr k) or (a.limbs[i+1] shl (wordBitWidth - k))
+  a.limbs[a.limbs.len-1] = a.limbs[a.limbs.len-1] shr k
+
+func dbl[T: DynWord](a: var BigNum[T]): bool =
+  ## In-place multiprecision double
+  ##   a -> 2a
+  var carry, sum: T
+  for i in 0 ..< a.limbs.len:
+    let ai = T(a.limbs[i])
+    addC(carry, sum, ai, ai, carry)
+    a.limbs[i] = sum
+
+  result = bool carry
+
+func csub[T: DynWord](a: var BigNum[T], b: BigNum[T], ctl: bool): bool =
+  ## In-place optional substraction
+  ##
+  ## It is NOT constant-time and is intended
+  ## only for compile-time precomputation
+  ## of non-secret data.
+  var borrow, diff: T
+  for i in 0 ..< a.limbs.len:
+    let ai = T(a.limbs[i])
+    let bi = T(b.limbs[i])
+    subB(borrow, diff, ai, bi, borrow)
+    if ctl:
+      a.limbs[i] = diff
+
+  result = bool borrow
```

Comment:
The reason I did not use those is that they use statically sized arrays. It would require us to assign a static argument to many of the related procs that end up calling any of the above by needing any of the globals.